🚀 2단계 - 문자열 계산기

rima361 opened the pull request.

Commit:
feat: 문자열 계산기의 입력값 확인하는 스펙
- null, 빈 공백이거나 사칙연산 기호가 아닌지 확인

Commit:
feat: 문자열 계산기 사칙연산 스펙
- 덧셈,뺄셈,나눗셈,곱셈

Commit:
feat: 문자열 계산기 사칙연산 스펙
- 덧셈

Commit:
feat: 문자열 계산기 입력값에서 숫자와 연산자를 분류하는 기능 추가

Commit:
feat: 문자열 계산기 입력값에서 숫자와 연산자를 분류하는 기능 추가
- 연산자 모델 추가

Commit:
feat: 문자열 계산기 입력값 받는 기능 추가

Commit:
feat: 문자열 계산기 계산 스펙

Commit:
refactor: 문자열에서 숫자 / 사칙연산 기호 추출하는 행위는 각 객체에게 위임한다.
- 1단계 코드 리뷰 반영

Commit:
refactor: Number 의 멤버변수를 외부에 노출하지 않도록 리팩토링

jayden-17 commented on src/main/java/calculator/view/CalculateInfo.java:
```diff
@@ -0,0 +1,42 @@
+import calculator.Operation;
+import calculator.string.StringUtils;
+
+public class CalculateInfo {
```
CalculateInfo 는 이름이 모호한 것 같습니다.
이 객체를 더 잘 표현할 수 있는 이름은 없을까요 ?

rima361 replied:
`1+1, 2*2` 요런걸 `수식(Numbers and Operations)` 이라고 부르나봐요..?
그래서 MathExpression 로 수정할까 합니다

jayden-17 commented on src/main/java/calculator/Operation.java:
```diff
@@ -0,0 +1,52 @@
+    MINUS("-", MinusCalculator.instance),
+    MULTIPLY("*", MultiplyCalculator.instance),
+    DIVIDE("/", DivideCalculator.instance);
+
```
구현해야할 메소드가 1개만 있을 경우 람다로 표현할 수 있습니다.

jayden-17 commented on src/main/java/calculator/Number.java:
```diff
@@ -0,0 +1,69 @@
+
+import calculator.string.StringUtils;
+
+public class Number {
```
자주 쓰이는 ZERO 상수화 

Number 객체의 책임과 역할을 잘 나눈 것 같습니다 👍

jayden-17 commented on src/main/java/calculator/Number.java:
```diff
@@ -0,0 +1,69 @@
+
+    @Override
+    public String toString() {
+        if(value == null) {
```
value 가 null 일 수 있는건가요 ? 
특별한 이유가 없다면 Number 이 생성될 때 막는게 어떨까요 ?
허용 되는 순간 항상 value 가 없음을 확인해야 합니다.

jayden-17 commented on src/main/java/calculator/view/CalculateInfo.java:
```diff
@@ -0,0 +1,42 @@
+    }
+
+    public static CalculateInfo of(final String value) {
+        if (!new CalculatorInputValidator(value).isValid()) {
```
어떤 의도로 CalculatorInputValidator 를 별도의 객체로 분리했을까요?

rima361 replied:
사용자 입력값 유효성 검사에 대한 테스트 코드가 필요하다고 생각이 들었는데,
입력 모델에 책임을 주자니 private 로 접근을 제한하는 구조로 만들어져서 분리했어요~

private를 메서드를 테스트하고 싶을때 객체를 역할에 맞게 더 잘게 쪼개면 된다고 생각했고,
CalculatorInputValidator 는 유효성을 확인하는 역할을 가지도록 구현했어요.

jayden-17 commented on src/main/java/calculator/view/CalculateInfo.java:
```diff
@@ -0,0 +1,42 @@
+        if (!new CalculatorInputValidator(value).isValid()) {
+            throw new IllegalArgumentException();
+        }
+        return new CalculateInfo(Number.extractNumber(value), Operation.extractOperation(value));
```
input 에서 모든 숫자를 추출하는게, input 에서 모든 operation 을 추출하는게 
Number 와 Operation 의 책임으로 적절해 보이지 않습니다.

rima361 replied:
흠 input 의 책임으로 가지는게 좋을까요,, 한 객체의 책임이 많아질수록 테스트 코드가 짜기가 어려워지네요.

jayden-17 reviewed (changes requested):
`"+ * / 2 3 4 2:10"`, `"+ * / * + 2 3 4 2:10"`이런 수식도 동작할 것 같아요 ~ 

현재 number 와 operation 의 배치에 대한 유효성 검사가 없습니다. 
operation 이 남을 경우에도 마찬가지 입니다.

Commit:
refactor: 코드 리뷰 반영
- CalculateInfo class 명 변경 (MathExpression)
- Operation 의 calculator 를  람다식으로 변경
- Number 객체 생성시 null 체크 추가 및 toString null 체크 로직 제거
- CalculateInfo 로 Number/Operator 추출 책임 이동

Commit:
test: 연산자로 끝나는 문자열 테스트 케이스 추가

jayden-17 commented on src/main/java/calculator/Number.java:
```diff
@@ -0,0 +1,67 @@
+        return new Number(value / target.value);
+    }
+
+    public boolean isOdd() {
```
이 부분도 고민해볼만한 영역인데요 ~ 

Number 의 책임이 계속 늘어날 때는 적절한 역할을 하고 있는지 한번 확인해보면 좋을 것 같아요 ~ 

isOdd 는 사실 수식 검증을 위한 index 의 포지션 확인 용도로 사용 되고 있어요 ~ 

기존 Number의 책임인 연산을 위한 숫자와는 차이가 있어 보여요 ~

jayden-17 commented on src/main/java/calculator/StringCalculator.java:
```diff
@@ -0,0 +1,18 @@
+
+import calculator.view.MathExpression;
+
+public class StringCalculator {
```
이 부분 조금만 더 고민해서 정리해보면 좋을 것 같아요 ~ 

- MathExpression 자체가 calculate 를 할 수 있어 보임. 
  - target.getNumbers() 와 target.pollOperation() 가 결국 내부 구현을 외부로 노출한 걸로 보입니다. 
- 패키지 구조상 MathExpression 은 뷰 모델이고, StringCalculator 는 도메인 모델로 보이는데 도메인이 뷰에 의존하고 있음.
  - StringCalculator 에서 MathExpression 를 인자로 받고 있음

jayden-17 commented on src/main/java/calculator/string/StringUtils.java:
```diff
@@ -16,4 +19,9 @@ public static String[] split(final String value) {
         return value.split(DEFAULT_SPLITTER);
     }
+
+    public static char[] getChars(final String value) {
```
객체지향 교육 목적이니 최대한 `~Utils` 는 사용하지 않기로 하고 
구체적인 객체를 만들거나 사용 되는 객체에 해당 기능을 옮겨 보는건 어떨까요?

`getChars` 의 경우에는 중간 공백을 모두 제거한 후 char 로 변환하고 있는데, 
어떠한 경우에 이 기능이 필요한지 
이 기능은 어떤 객체가 알고 있어야 할지 ? 조금 더 고민해보면 어떨까요?

나중에 더 많은 요구사항이 추가되고 그때 동일한 요구사항이 생겼을 경우 공통 로직으로 추출해볼 수 있을 것 같아요 ~ 
`extractNonEmptyChars()` 와 같이 더 구체적인 이름으로 분리해보면 좋을 것 같아요 ~
(`getChats()` 도 string 에서 문자를 가져오려는 의도로 네이밍 한게 맞죠 ? 😄 )

jayden-17 commented on src/main/java/calculator/view/CalculatorInputValidator.java:
```diff
@@ -0,0 +1,42 @@
+import calculator.Operation;
+import calculator.string.StringUtils;
+
+public class CalculatorInputValidator {
```
이 부분은 제가 설계의 의도를 잘 파악하지 못해서 그럴 수 있을 것 같아요 ~ 
일단 제 의견을 남겨볼게요, 같이 논의해보면 좋을 것 같아요 ~ 

위에서도 언급했는데 
`StringCalculator`, `MathExpression`, `CalculatorInputValidator` 의 
책임과 관계가 명확하게 보이질 않습니다. 

작업한 코드 기반으로 조금 배치만 다시해보면 다음과 같이도 표현할 수 있을것 같아요 

`CalcaulatorInput calcaulatorInput = new CalcaulatorInput(input);`

```
public class StringCalculator {

    public static Number calculate(final CalcaulatorInput calcaulatorInput) {
        MathExpression mathExpression = calcaulatorInput.mathExpression();
        return mathExpression.calculate().orElse(ZERO)l
    }
```

jayden-17 reviewed (changes requested):
피드백 남겨놨어요 ~ 부디 도움이 되는 피드백이면 좋겠네요 🙏 

아 그리고 한가지 더 요청 드릴게 있습니다.

README.md 에 요구사항을 분석해서 작성하고 어떤 기능을 추가할지 작성해보면 좋을 것 같아요 ~ 
어떤 기능을 추가할지 정리한 것들이 자연스럽게 단위테스트 단위가 되면 TDD 로 진행하는데 훨씬 더 도움이 되 않을까요 ?

경험상 처음부터 단위테스트 도출과 코드 작업이 완벽하게 연결되지는 않았어요, 두번째 요구사항을 구현하면서 첫번째 테스트가 깨지거나 구현이 전반적으로 변경된 적도 있었어요 ~ 
다만 습관적으로 연습하다 보면 내가 도출한 기능들이 단위테스트로 자연스럽게 연결될 때도 있었습니다. 
예전에 이런식으로 의도적으로 작성해봤는데 도움이 많이 되었어요 ~ 
https://github.com/next-step/java-racingcar/blob/jinuk17/README.md

그래서 하는 말인데 ㅋㅋㅋ
`입력 문자열의 숫자와 사칙 연산 사이에는 반드시 빈 공백 문자열이 있다고 가정한다.` 라고 있었는데 
공백을 다 없애기 보단 공백으로 `split` 처리해서 1번은 숫자, 2번은 연산자 이렇게 판단할 수도 있습니다.
현재 버전에서는 두자리 숫자는 처리가 안되겠죠 ? ㅠ char 로 취급해서 ㅠ
(ex, 11 + 22)
